src/XrdCeph/XrdCephBulkAioRead.hh

```diff
@@ -0,0 +1,51 @@
+#include <string>
+#include <map>
+#include <list>
+#include <tuple>
+#include <rados/librados.hpp>
+
+
+class bulkAioRead {
+  typedef std::tuple<ceph::bufferlist*, char*, int*> ReadOpData;
+  typedef void (*logfunc_pointer) (char *, ...);
+
+  /*
+ * Class is used to execute read operations against rados striper files *without* usage of rados striper.
+ * Reads are based on ceph read operations.
+ *
+ * The interface is similar to the one that ceph's read operation objects has:
+ * 1. Instantiate the object.
+ * 2. Declare read operations using 'read' method, providing the output buffers, offset and length.
+ * 3. Submitn operation and wait for results using 'submit_and_wait_for_complete' method.
+ * 4. Copy results to buffers with 'get_results' method. 
+  */ 
+  public:
+  bulkAioRead(librados::IoCtx *ct, logfunc_pointer ptr, std::string filename, size_t object_size);
+  ~bulkAioRead();
+
+  void clear();
+  int submit_and_wait_for_complete();
+  ssize_t get_results();
+  int read(void *out_buf, size_t size, off64_t offset);
+
+  private:
+  int addRequest(size_t obj_idx, char *out_buf, size_t size, off64_t offset);
+  librados::IoCtx* context;
+  std::list<ReadOpData> buffers;
+  /* [ 
+ *     (<ceph_bufferlist>, <client_buffer>, <retval_pointer>),
+ *     ...
+ *   ]
+ * */
+
+  std::map<size_t, std::pair<librados::ObjectReadOperation*, librados::AioCompletion*>> operations;
```

Comment:
vector maybe then?

Reply:
Developing several points, you could replace the pair with (say):
```
struct OpData {
  librados::ObjectReadOperation rd_op;
  CmplPtr cmpl;
  std::string obj_name;
};
```
(or the equivalent tuple, but I think a `struct` is more readable, especially at access).
Declare `CmplPtr` as:
```
class CmplPtr {
  librados::AioCompletion *ptr;
public:
  CmplPtr() : ptr(librados::Rados::aio_create_completion()) {
    if (ptr == NULL) throw std::bad_alloc();
  }
  ~CmplPtr() { ptr->release(); }
  librados::AioCompletion *get() const { return ptr; }
};
```
(or you could use a `std::unique_ptr` with a custom deleter, but I think it's as much work).
Then in `addRequest`, you can use the empty `obj_name` to detect a new addition to the map, and compute the name then:
```
auto &op_data = operations[obj_idx];
if (op_data.obj_name.empty()) {
  // It's a new entry.
  char obj_suffix[18];
  ...
  op_data.obj_name = ...
}
```
Similarly with `ReadOpData`:
```
struct ReadOpData {
  ceph::bufferlist bl;
  char *out_buf;
  int rc;
  ReadOpData(char *out_buf) : out_buf(out_buf) { }
};
```
Back in `addRequest`:
```
ReadOpData &last = buffers.emplace_back(out_buf);
op_data.rd_op.read(offset, size, &last.bl, &last.rc);
```
`get_results` should iterate `buffers` with a reference.
`bulkAioRead::clear()` becomes just a couple of `clear()` calls, and you can reduce the error checking in `read` to just an all-encompassing `try` catching `bad_alloc`.
For `wait_for_complete`, a `void` return should be possible again.

Reply:
Good point, thanks!
That was implemented (see updates).

The only thing i'd say is that it seems like ther is no need to check whether the operation is new or not, so no need to keep `obj_name ` in `OpData`.

Reply:
Hmm, perhaps I badly stated the intent of the empty-string test; it's not necessary to know that the operation is new, just that the entry is new, requiring `obj_name` to be filled in.  Just thought it would be a nice way to get the allocations done before initiating the operations, and so avoid initiating an operation that you later abandon because the name allocation for a subsequently initiated operation failed.
An alternative would be to have one loop at the start of `submit_and_wait_for_complete` to allocate all the names, then a separate loop to initiate all the operations.  But then you'd need a map to preserve the names from one loop to the other.  But then you've already got the `operations` map, so you may as well bung it in there.
Anyway, it's fine as it is.